### ISSUE TITLE
Project 2 Peer Review thread

### ISSUE THREAD

**Issue by susanBuck** [susanBuck/e15-spring23#21]
In this week’s upcoming assignment (Week 9, Mar 28), I'll prompt you to complete a peer review of a classmate’s Project 2. You will share your completed review in a reply to this thread.

Full instructions of the peer review process are in the Week 9 assignment.

[Peer review spreadsheet...](https://docs.google.com/spreadsheets/d/1gasdelov7F7NEXtUeeAtygt5wRFAsIWZgtP7SrVBKEY/edit#gid=0)

**Comment by lmarshman** [comment]
# Peer Review for dannybodner21

1. What is the Github URL of the project you are reviewing?
https://github.com/dannybodner21/e15/tree/main/p2

2. What is the production URL of the project you are reviewing?
http://e15p2.dannybodner.com/

3. Are you able to produce any errors or unexpected results when using the site?
No.

4. Is there appropriate separation of concerns amongst controllers and views?
Yes. 

5. Are there any parts of the code that you found interesting or taught you something new?
I’m interested in the logic behind the form display. There is a dropdown menu that displays up to three versions of the form, allowing the user to enter different cryptocurrencies/values. It’s not immediately obvious to me what’s controlling the form display. I would have thought there would be an if statement to control how many versions of the form are displayed, but this doesn’t seem to be the case. I’m assuming it must relate to the inclusion of: <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.3/jquery.min.js"></script>

6. Are there any parts of the code that you don't understand?
My answer is essentially the same as the previous question. I don’t understand what’s controlling the form display, but I’d like to!

7. Are there any parts of the code that you feel could be written in a more efficient or clearer manner?
The code seems to be written efficiently to me.

8. Are there aspects of the code that you feel were not self-evident and would benefit from comments?
I think overall the code is commented appropriately, without being verbose.

9. Are there any best practices discussed in course material that you feel were not addressed in the code?
All best practices seem to have been addressed.

10. Do you have any additional comments not covered in the above questions?
No additional comments.

**Comment by dannybodner21** [comment]
## Peer Review for ROK862

1. What is the Github URL of the project you are reviewing? 
https://github.com/ROK862/e15/tree/main/p2

2. What is the production URL of the project you are reviewing? 
http://e15p2.appsuits.org/

3. Are you able to produce any errors or unexpected results when using the site? 
It lets you input a date in the future, and then reloads the home page without any output. The user should be informed of the issue. After looking at the code, I noticed that there is code in place for if an invalid date is input, but it doesn’t seem to do what it is supposed to. It looks like when an invalid date is input it sends the user back to the home page with an error variable set to true and a message to display the error, but on the home page view there is nowhere for this to be displayed. If I am reading this correctly, the error message code block might be missing.

4. Is there appropriate separation of concerns amongst controllers and views? E.g. is there any non-display specific logic in views, or display-specific code in controllers? 
Yes, the views are easy to read through and contain all the display specific logic. The controller functionality is clean, and contains comments throughout making it easy to follow.

5. Are there any parts of the code that you found interesting or taught you something new?
I like the use of the ApiProvider to access all the stock data.

6. Are there any parts of the code that you don't understand? 
I have a loose grasp of exactly what the API call is doing in the ApiProvider Controller.

7. Are there any parts of the code that you feel could be written in a more efficient or clearer manner? 
No, I feel like the code is pretty concise. It gets the job done in few lines of code.

8. Are there aspects of the code that you feel were not self-evident and would benefit from comments? 
Comments were used throughout most of the controller doing the main logic, but if you don't know much about API calls you might get confused.

9. Are there any best practices discussed in course material that you feel were not addressed in the code? 
I believe the best practices discussed in class were demonstrated appropriately.

10. Do you have any additional comments not covered in the above questions?
Good work. I like how the project displays the data in question but also additional data about the given stock. Personally, I think on the results page you should add in units. Right now it shows the profits/loss as a number without any unit, such as dollars. Overall, I like the project.

**Comment by dannybodner21** [comment]
@lmarshman 

To control the form display I am using JavaScript and the help of that JQuery line you mentioned. At the bottom of my index.blade.php file there is a script tag that controls what the user is seeing based on the changing of the dropdown input.

**Comment by bar181** [comment]
### Peer Review for traveler9878

What is the Github URL of the project you are reviewing?
[https://github.com/traveler9878/e15/tree/main/p2](https://github.com/traveler9878/e15/tree/main/p2)

What is the production URL of the project you are reviewing?
[http://e15p2.davidcurtis.me/](http://e15p2.davidcurtis.me/)

Are you able to produce any errors or unexpected results when using the site?
- The radio button for stars will default to star 3 (if you select stars 4 or 5).  In your home.blade the radio button for 3star is always checked meaning the same radio group before 3star will not work but ones after will work 
- Minor: you reference app.css in main.blade.php but the file does not exist resulting in a 404 error in the console.  I see the css file in resources but it should be in the public/css folder.
- Your radio button for active includes a condition that will always return checked.  Consider having an OR condition to check if old is checked or old does not exist.  This would cause an error if you reorder the radio buttons and have the YES option last.


Is there appropriate separation of concerns amongst controllers and views? E.g. is there any non-display specific logic in views, or display-specific code in controllers?
- Your NoteController should return the session data as part of the view.  This would eliminate the need to use php code to access the $_SESSION data in your read.blade. 

Are there any parts of the code that you found interesting or taught you something new?
- Looks good

Are there any parts of the code that you don't understand?
- Straight forward and everything makes sense.

Are there any parts of the code that you feel could be written in a more efficient or clearer manner?
- moving the session data would clean up the read.blade. Other than that all looks good. 

Are there aspects of the code that you feel were not self-evident and would benefit from comments?
- All good ! 
- Your well labeled input fields and clean code eliminates the need for comments.

Are there any best practices discussed in course material that you feel were not addressed in the code?
- The style tag in your layouts/main should be moved to the app.css file

Do you have any additional comments not covered in the above questions?
- delete unused files such as welcome.blade 
- your labels for the star ratings all reference '1star'
- your old content for the textarea always adds several spaces, I had the same issue and found if you have the old immediately after the end of the textarea tag it works (i.e. do not put on the next line)
- Missing end body tag in layouts/main.blade 
- Overall, Like it can can't wait to see project 3 !

**Comment by andlearnmore** [comment]
### Peer review for bar181
**What is the Github URL of the project you are reviewing?** 
[https://github.com/bar181/e15/tree/main/p2](url)

**What is the production URL of the project you are reviewing?** 
[e15p2.bradross.me](url)

**Are you able to produce any errors or unexpected results when using the site?** 
No, I'm not able to produce errors or unexpected results. The form submits as expected and shows the input both when I submit it _without_ all fields complete and after I submit it successfully. The validation is working consistently, with appropriate explanations. From the code, I noticed additional validation that I hadn’t tested originally (title must be between 3-15 characters; the message must be at least 3 characters), which also worked as expected.

**Is there appropriate separation of concerns amongst controllers and views? E.g. is there any non-display specific logic in views, or display-specific code in controllers?** 
Yes - it is all organized very well with good separation of concerns.

**Are there any parts of the code that you found interesting or taught you something new?** 
I thought this code was interesting:
`'image' => self::getSrc($formData['imageType'])`
And learned from how you used a method within the controller to get the image. This is something I’d like to practice doing.

**Are there any parts of the code that you don't understand?** 
No, there’s nothing that’s unclear at all.

**Are there any parts of the code that you feel could be written in a more efficient or clearer manner?** 
I wish I had more constructive feedback, but everything was very clear and easy to follow!

**Are there aspects of the code that you feel were not self-evident and would benefit from comments?** 
This project was very well-commented. It was written very neatly and was easy to read.

**Are there any best practices discussed in course material that you feel were not addressed in the code?** 
The project took advantage of things we’ve covered in lecture, including making use of the ternary operator when determining how the form fields should display, clear indentation and organization. 

**Do you have any additional comments not covered in the above questions?**
I always learn something from your code and projects, Brad!

**Comment by ghost** [comment]
### **Peer Review for sicwere**

What is the Github URL of the project you are reviewing?
<https://github.com/sicwere/e15/tree/main/p2>

What is the production URL of the project you are reviewing?
<http://e15p2.sicwere.online/>

Are you able to produce any errors or unexpected results when using the site?
No, I am not able to produce errors or unexpected results when using the site.

Is there appropriate separation of concerns amongst controllers and views?
E.g. is there any non-display specific logic in views, or display-specific code in
controllers?
Yes, there is an appropriate separation of concerns amongst controllers and views.

Are there any parts of the code that you found interesting or taught you
something new?
The formula and the method for calculating BMI is very interesting and informative.  

Are there any parts of the code that you don't understand?
Yes.

```php
foreach($ranges as $key => $value) { 
if($bmi < $value) { 
$category = $key; 
break; 
```
The method of iterating through the ranges array using foreach loop and break is very interesting. However, what if the BMI variable is greater than the 'overweight' => 29.9 value.

Are there any parts of the code that you feel could be written in a more efficient
or clearer manner?
No, the code was written efficiently.

Are there aspects of the code that you feel were not self-evident and would
benefit from comments?
No, the code is clear and precise.

Are there any best practices discussed in course material that you feel were not
addressed in the code?
The code follows all the best practices discussed in the course material.

Do you have any additional comments not covered in the above questions?
It would have been a nice feature to see a user’s inputted values along with the result of the BMI output. Overall, the code is efficient and presented neatly. The functions used are concise and adequate. The work is very well done and follows all the best practices presented in the course.

**Comment by sicwere** [comment]
**What is the Github URL of the project you are reviewing?**

https://github.com/lmarshman/e15/tree/main/p2

**What is the production URL of the project you are reviewing?**

http://p2.lauramarshman.com/

**Are you able to produce any errors or unexpected results when using the site?**

Yes. The tsp-to-tsp conversion does not work. If I enter a 1 in the Amount textbox, the conversion is not 1, but 0.16666666666667 tsp. Higher values are scaled proportionally; if 2 is entered, the conversion is 0.33333333333333 tsp instead of 2, for example. This error is not present in any other conversion.

In addition, the tbs-to-oz conversion crashes with a "Undefined variable $conversion" error.

Also, not an error as such, but on the Welcome page, measurements is misspelled as "measuremnts".

**Is there appropriate separation of concerns amongst controllers and views?**

Yes.

**Are there any parts of the code that you found interesting or taught you something new?**

Not particularly.

**Are there any parts of the code that you don't understand?**

I'm not sure what's causing the errors I mentioned above.

**Are there any parts of the code that you feel could be written in a more efficient or clearer manner?**

Aside from fixing the errors mentioned above, I do think that the FormController could have been made a bit more efficient. For example, if $convert1 == $convert2, then the amount will be unchanged and the program does not need to run the conversion code. I would have written this:

```php
# If the selected measurements are the same, returns the original amount.

if ($convert1 == $convert2) {
     $conversion = $amount;
}
else {
# conversion code
}
```
I would have done the same for the conversion code:

```php
if(($convert1 == 'tsp') {

}
else if(($convert1 == 'tbs') {

}

etc.

```


**Are there aspects of the code that you feel were not self-evident and would benefit from comments?**

No.

**Are there any best practices discussed in course material that you feel were not addressed in the code?**

No.

**Do you have any additional comments not covered in the above questions?**

I think it would have nice to have an "Unchanged amount" radio button to go with the Halve, Double, and Triple options, or otherwise have a way to clear a selected radio button. If I select Halve, Double, and Triple the Recipe, the only way to clear it is to refresh the page.

**Comment by coeure-d-elephante** [comment]
**What is the Github URL of the project you are reviewing?**

[Github url](https://github.com/jacksuwa/e15/tree/main/p2)



**What is the production URL of the project you are reviewing?**
[production URL](http://e15p2.jacksuwa.online/)
Are you able to produce any errors or unexpected results when using the site?

No
**Is there appropriate separation of concerns amongst controllers and views?** 
 
Yes

**Is there any non-display specific logic in views, or display-specific code in controllers?**
No

**Are there any parts of the code that you found interesting or taught you something new?**
```php 
 <input type='radio' name='status' id='single' value='single'
                {{ old('status') == 'single' ? 'checked' : '' }}>
```
Similarly to the dilemma I faced with my own project. I can see how this person used the old() helper method in this radio input. This is adding to the repetitions that I'm seeing as to how this method can be best used


**Are there any parts of the code that you don't understand?**

The code is pretty easy to follow and understand. My question is more in the design. Please see following q's.


**Are there any parts of the code that you feel could be written in a more efficient or clearer manner?**
I really like the concept of your project Jackson! 

However, I am a bit thrown off by the "status" input of "single" or "married". I think it would make sense for each radio option to have its own results. That if someone selects "single" the result would be their wage as a single person. And that if someone is "married" then the result and calculation would be different.

**Are there aspects of the code that you feel were not self-evident and would benefit from comments?**

I would have liked to see logic to accommodate my comment above. I can see that there is only 1 calculation for both "single" and "married" options right now. And that the result doesn't change for either "single" or "married". The only variable that changes is the indication of the radio input. 

```php

  if ($request->has('deductTax')) {
            $percentage = 22;
            $wage = $wage - ($percentage / 100) * $wage;
        }
```

**Are there any best practices discussed in course material that you feel were not addressed in the code?**
No. 

**Do you have any additional comments not covered in the above questions?**

@jacksuwa! Apart from the concerns I mentioned above, I really liked the concept of your project 2! Well done! 😊

**Comment by ROK862** [comment]
## Peer review for Jacksuwa

- What is the Github URL of the project you are reviewing?
[Github URL](https://github.com/jacksuwa/e15/tree/main/p2)

- What is the production URL of the project you are reviewing?
[Product URL](http://e15p2.bradross.me/)

- Are you able to produce any errors or unexpected results when using the site?
I was able to reproduce an error by submitting the postcard form and refreshing the page. We then get a console error as such:

Failed to load resource: the server responded with a status of 520 ()
tailwind.min.css:1

The person may be trying to reference an external CSS file that returns an unknown, empty, or unexpected response.

```HTML
<!-- REF: --> 
<link href="https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css" rel="stylesheet"> 
<!-- line 8, views/layouts/main.blade.php -->
```

- Is there an appropriate separation of concerns amongst controllers and views? E.g. is there any non-display specific logic in views, or display-specific code in controllers?

I would abstract the contents of the getSrc function to config.php such that if the application grows, we could access the values from anywhere as per the need. If that's the case, we would not need the getSrc function at all.

```php
// REF:  
$images = [...]; 
// line 66, PageController.php.
```
Are there any parts of the code that you found interesting or taught you something new?
The constraint on the getSrc function's return type string. I never tried using that in PHP. Really cool!

REF:  public static function getSrc($item): string line 64, PageController.php.

Also, let's not forget the conditional return; I know we covered this in lecture notes, but it shows care in implementation.

```php
// REF:  
return $images[$item] ?? $images['cute']; 
// line 73, PageController.php.
```

- Are there any parts of the code that you don't understand?

No, it was a simple implementation.
Are there any parts of the code that you feel could be written in a more efficient or clearer manner?
In terms of functional fulfillment, I would have reconsidered the "create" function in the page controller. With the current implementation, there is just no way to create more than one additional postcard.

For example, the session variable newPostcard is an object literal which houses an array of properties defining a single object--postcard in this case. This means, when you create a new/additional postcard, you simply replace the newPostcard object literal with a new value or assign it at the very least.

It would be better to have the sessions' "newPostcard" variable operate as a sequential array instead. Then rendering could be done with a for loop in the associated view. 

```php
// REF:  
public function create(Request $request) 
// line 38, PageController.php.
```


- Are there aspects of the code that you feel were not self-evident and would benefit from comments?

No, the code is well-commented. I believe this person did a great job in that regard.
Are there any best practices discussed in course material that you feel were not addressed in the code?
It seems okay. I'm just considering drying the code up. 

- Do you have any additional comments not covered in the above questions?

I think the code is well written apart from aspects of incorrect implementation of the create function.

**Comment by traveler9878** [comment]
Peer review for AndLearnMore:

Your Answer:
What is the Github URL of the project you are reviewing?
https://github.com/andlearnmore/e15/tree/main/p2

What is the production URL of the project you are reviewing?
http://e15p2.andlearn.me/planner

Are you able to produce any errors or unexpected results when using the site?
The form processing page just dumps the results.  Form page looks good.

Is there appropriate separation of concerns amongst controllers and views? E.g. is there any non-display specific logic in views, or display-specific code in controllers?
I didn't seen any issues

Are there any parts of the code that you found interesting or taught you something new?
I thought it was interesting that they incorporated json data

Are there any parts of the code that you don't understand?
No

Are there any parts of the code that you feel could be written in a more efficient or clearer manner?
It seems pretty clean, maybe a little linear but I'm not introducing classes at this stage would have been appropriate for the assignment, but classes could be effective in breaking up some of the code in my opinion.

Are there aspects of the code that you feel were not self-evident and would benefit from comments?
I never trust comments actually, 'they lie', when people update the code and don't update the comments

Are there any best practices discussed in course material that you feel were not addressed in the code?
No

Do you have any additional comments not covered in the above questions?
I thought it was good work, they missed the answer to my forum post about just dumping the form post results on the action page.
